### PR TITLE
Replace resource ids when anonymization

### PR DESCRIPTION
**Pull request by moria97**
1. Replace resource ids
2. Replace Reference.reference by extract resource id from possible uuid/oid/url string
3. Save id mapping file for possible re-Id purpose

### CODE REVIEW

File: src/Fhir.Anonymizer.Core.UnitTests/Resource/ResourceIdTransfomerTests.cs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+using System.Linq;
+using Fhir.Anonymizer.Core.Resource;
+using Hl7.Fhir.ElementModel;
+using Hl7.Fhir.Serialization;
+using Hl7.FhirPath;
+using Xunit;
+
+namespace Fhir.Anonymizer.Core.UnitTests
+{
+    public class ResourceIdTransformerTests
```
Comment (BoyaWu10) [inline src/Fhir.Anonymizer.Core.UnitTests/Resource/ResourceIdTransfomerTests.cs:12]: Typo in file name "Transfomer"

File: src/Fhir.Anonymizer.Tool/Program.cs [hunk 2]
```diff
@@ -17,6 +18,8 @@ class Options
             public string OutputFolder { get; set; }
             [Option('c', "configFile", Required = false, Default = "configuration-sample.json", HelpText = "Anonymizer configuration file path.")]
             public string ConfigurationFilePath { get; set; }
+            [Option('m', "mappingFile", Required = false, Default = "mapping-ids.dat", HelpText = "Resource Id mapping file path.")]
+            public string MappingFilePath { get; set; }
```
Comment (BoyaWu10) [inline src/Fhir.Anonymizer.Tool/Program.cs:22]: This parameter is used for output? It seems we never load it.

Comment (BoyaWu10) [reply src/Fhir.Anonymizer.Tool/Program.cs:22]: Besides, it seems transforming id is always done by engine. It is not optional right? And do we need the same logic for adf?

Comment (moria97) [reply src/Fhir.Anonymizer.Tool/Program.cs:22]: Yes. This file option is for saving mapping for re-Id purpose. Should be add in adf too

Comment (BoyaWu10) [reply src/Fhir.Anonymizer.Tool/Program.cs:22]: Got it. If it is used for output, can we add some explanation like "Resource id mapping file saved from id transformation"?

File: src/Fhir.Anonymizer.Core/Resource/ResourceIdTransformer.cs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
...MedicationStatement|MedicinalProduct|MedicinalProductAuthorization|MedicinalProductContraindication|MedicinalProductIndication|MedicinalProductIngredient|MedicinalProductInteraction|MedicinalProductManufactured|MedicinalProductPackaged|MedicinalProductPharmaceutical|MedicinalProductUndesirableEffect|MessageDefinition|MessageHeader|MolecularSequence|NamingSystem|NutritionOrder|Observation|ObservationDefinition|OperationDefinition|OperationOutcome|Organization|OrganizationAffiliation|Patient|PaymentNotice|PaymentReconciliation|Person|PlanDefinition|Practitioner|PractitionerRole|Procedure|Provenance|Questionnaire|QuestionnaireResponse|RelatedPerson|RequestGroup|ResearchDefinition|ResearchElementDefinition|ResearchStudy|ResearchSubject|RiskAssessment|RiskEvidenceSynthesis|Schedule|SearchParameter|ServiceRequest|Slot|Specimen|SpecimenDefinition|StructureDefinition|StructureMap|Subscription|Substance|SubstanceNucleicAcid|SubstancePolymer|SubstanceProtein|SubstanceReferenceInformation|SubstanceSourceMaterial|SubstanceSpecification|SupplyDelivery|SupplyRequest|Task|TerminologyCapabilities|TestReport|TestScript|ValueSet|VerificationResult|VisionPrescription)\/(?<id>[A-Za-z0-9\-\.]{1,64})"),
```
[the remark this answers is not in the document]
Comment (skywalker-wu) [reply src/Fhir.Anonymizer.Core/Resource/ResourceIdTransformer.cs:22]: +1

File: src/Fhir.Anonymizer.Core/AnonymizerEngine.cs [hunk 4]
```diff
@@ -47,17 +48,20 @@ public string AnonymizeJson(string json, bool isPrettyOutput = false)
                 throw new Exception("Failed to parse json resource, please check the json content.", innerException);
             }
 
+            var anonymizedNode = AnonymizeResourceNode(root);
+            ResourceIdTransformer.Transform(anonymizedNode);
```
Comment (skywalker-wu) [inline src/Fhir.Anonymizer.Core/AnonymizerEngine.cs:52]: do we need option to enable/disable this?

File: src/Fhir.Anonymizer.Core/Resource/ResourceIdTransformer.cs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            }
+        }
+
+        public static void SaveMappingFile(string mappingFile)
+        {
+            using var fileStream = new FileStream(mappingFile, FileMode.Create);
+            using var writer = new StreamWriter(fileStream);
+            foreach(var k in _resourceIdMap.Keys)
+            {
+                writer.WriteLine($"{k}{MappingFileDelimiter}{_resourceIdMap[k]}");
```
Comment (skywalker-wu) [inline src/Fhir.Anonymizer.Core/Resource/ResourceIdTransformer.cs:64]: Can we just write json string of dictionary?

Comment (skywalker-wu) [reply src/Fhir.Anonymizer.Core/Resource/ResourceIdTransformer.cs:64]: Then we can avoid parse it ourself.

File: src/Fhir.Anonymizer.Core/AnonymizerEngine.cs (hunk 4 above)
Comment (skywalker-wu) [reply src/Fhir.Anonymizer.Core/AnonymizerEngine.cs:52]: would it impact performance?

File: src/Fhir.Anonymizer.Core/Resource/ResourceIdTransformer.cs (hunk 3 above)
Comment (skywalker-wu) [inline src/Fhir.Anonymizer.Core/Resource/ResourceIdTransformer.cs:22]: Maybe put some sample in comments for better understanding?

**Comment by deepak-ms** [comment]
Let's wait for discussion/resolution of design document comments before submitting this change.

File: src/Fhir.Anonymizer.Core/Resource/ResourceIdTransformer.cs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+            new Regex(@"urn:oid:(?<id>[0-2](\.(0|[1-9][0-9]*))+)"),
+            new Regex(@"urn:uuid:(?<id>[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})")
+        };
+
+        private static readonly ConcurrentDictionary<string, string> _resourceIdMap = new ConcurrentDictionary<string, string>();
```
Comment (skywalker-wu) [inline src/Fhir.Anonymizer.Core/Resource/ResourceIdTransformer.cs:27]: not use static?

File: src/Fhir.Anonymizer.Core.UnitTests/Resource/ResourceIdTransfomerTests.cs (hunk 1 above)
Comment (moria97) [reply src/Fhir.Anonymizer.Core.UnitTests/Resource/ResourceIdTransfomerTests.cs]: Fixed, thanks!

File: src/Fhir.Anonymizer.Core/Resource/ResourceIdTransformer.cs (hunk 6 above)
Comment (moria97) [reply src/Fhir.Anonymizer.Core/Resource/ResourceIdTransformer.cs]: Thanks, changed ResourceIdTransformer to a dependency of engine for forward customization.

File: src/Fhir.Anonymizer.Core/Resource/ResourceIdTransformer.cs (hunk 3 above)
Comment (moria97) [reply src/Fhir.Anonymizer.Core/Resource/ResourceIdTransformer.cs]: Yes, have added the regex link defined in Hl7

File: src/Fhir.Anonymizer.Core/Resource/ResourceIdTransformer.cs (hunk 5 above)
Comment (moria97) [reply src/Fhir.Anonymizer.Core/Resource/ResourceIdTransformer.cs]: Changed to json serialization/deserialization, thanks

File: src/Fhir.Anonymizer.Core/AnonymizerEngine.cs (hunk 4 above)
Comment (moria97) [reply src/Fhir.Anonymizer.Core/AnonymizerEngine.cs]: And an option to enable this. And tested with some data, it makes no difference to the performance

File: src/Fhir.Anonymizer.Core/Resource/ResourceIdTransformer.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (moria97) [reply src/Fhir.Anonymizer.Core/Resource/ResourceIdTransformer.cs]: I believe so as it comes from Hl7.  Also attached the source of this definition in comments

**Pull request closed without merging**